godot-cell/src/lib.rs

```diff
@@ -0,0 +1,328 @@
+        // since we cannot have any other aliasing mutable references around to pass in.
+        unsafe { Ok(MutGuard::new(&self.get_ref().state, count, value)) }
+    }
+
+    /// Set the current mutable borrow as not aliasing any other references.
+    ///
+    /// Will error if there is no current possibly aliasing mutable borrow, or if there are any shared
+    /// references.
+    pub fn set_non_aliasing<'a, 'b>(
+        self: Pin<&'a Self>,
+        current_ref: &'b mut T,
+    ) -> Result<NonAliasingGuard<'b, T>, Box<dyn Error>>
+    where
+        'a: 'b,
+    {
+        let mut state = self.state.lock().unwrap();
+        let current_ptr = state.get_ptr();
+        let ptr = NonNull::from(current_ref);
+
+        if current_ptr != ptr {
+            // it is likely not unsound for this to happen, but it's unexpected
+            return Err("wrong reference passed in".into());
+        }
+
+        state.borrow_state.set_non_aliasing()?;
+        let old_ptr = state.get_ptr();
+        state.set_ptr(ptr);
+
+        Ok(NonAliasingGuard::new(&self.get_ref().state, old_ptr))
+    }
+
+    /// Returns `true` if there are any mutable or shared references, regardless of whether the mutable
+    /// references are aliasing or not.
+    ///
+    /// In particular this means that it is safe to destroy this cell and the value contained within, as no
+    /// references can exist that can reference this cell.
+    ///
+    /// Keep in mind that in multithreaded code it is still possible for this to return true, and then the
+    /// cell hands out a new borrow before it is destroyed. So we still need to ensure that this cannot
+    /// happen at the same time.
+    pub fn is_currently_bound(self: Pin<&Self>) -> bool {
+        let state = self.state.lock().unwrap();
+
+        state.borrow_state.shared_count() > 0 || state.borrow_state.mut_count() > 0
+    }
+}
+
+// SAFETY:
+// `T` is sync so we can return references to it on different threads.
+// Additionally all internal state is synchronized via a mutex, so we wont have race conditions when trying
+// to use it from multiple threads.
+unsafe impl<T: Sync> Sync for GdCell<T> {}
+
+/// Mutable state of the `GdCell`, bundled together to make it easier to avoid deadlocks when locking the
+/// mutex.
+#[derive(Debug)]
+struct CellState<T> {
+    /// Tracking the borrows this cell has. This ensures relevant invariants are upheld.
+    borrow_state: BorrowState,
+    /// Current pointer to the value.
+    ///
+    /// This is initialized upon first usage, as we cannot construct the cell pinned in general.
+    ///
+    /// When a reference is handed to a cell to enable re-entrancy, then this pointer is set to that
+    /// reference.
+    ///
+    /// We always generate new pointer based off of the reference currently in this field, to ensure any new
+    /// references are derived from the most recent `&mut` reference.
+    ptr: Option<NonNull<T>>,
```

Comment:
Well i know the pointer is non-null when initialized, so i made it a `NonNull`. Also because i use `NonNull` elsewhere. 

Covariance isn't an issue here since it's just used internally and there's an `UnsafeCell` higher up anyway that makes the `GdCell` invariant as well afaik.

Reply:
> Also because i use `NonNull` elsewhere.

Yes, but the rest of the library doesn't -- especially not `Option<NonNull<T>>` which is more idiomatically expressed as `*mut T`. Unless there's a very compelling reason, it would be good to use the simpler solution here.

I'll add a suggestion in the review, it significantly simplifies the `impl` block.

Reply:
I still believe it is better to use `NonNull<T>` here, for several reasons:
- There are cases where we use the bare `NonNull<T>`, such as in all the guards where we know they dont contain null pointers. 
- It better communicates intent, since we can specify where the pointer is assumed to be non-null. 
- It makes it easier to avoid accidentally dereferencing a null pointer, since the rust type system ensures that it is impossible to do so when using `NonNull<T>` (at least without `unsafe` which i do not use here for that). 

Also `Option<NonNull<T>>` has the same layout as `*mut T` so there should be no performance difference.

I dont think the docs are really saying to prefer `*mut T`, it's just that it has additional requirements over `*mut T` like `&mut T` does. Like you wouldn't recommend using `*mut T` over `Option<&mut T>` even though the same argument applies there. It is more dangerous to use `&mut T` than `*mut T` because the reference has additional properties over the raw pointer. And dont forget the first part of the paragraph you quoted:

> **This is often the correct thing to use when building data structures using raw pointers**, but is ultimately more dangerous to use because of its additional properties. If you’re not sure if you should use `NonNull<T>`, just use `*mut T`!

I am confident that the additional properties do not pose any danger here, and that the data structure i've created is correct with them, therefore the paragraph's warning does not apply. The warning only applies if we are unsure if the additional properties are correct for the use-case.

Reply:
My main concern are not the additional requirements (or lack thereof), more code consistency. The points you mention apply to a lot of the rest of gdext. But I deliberately chose against using `Option<NonNull<T>>`, because it often causes extra verbosity (double generic nested) without real safety in practice.

I showed how the implementation could look with `*mut T` -- to me it's not clear where we lose safety or intent through that. The methods can still return `NonNull<T>` where needed.

Reply:
> The points you mention apply to a lot of the rest of gdext.

Not as nicely, in the rest of gdext we do a lot of ffi which makes it a lot harder to confidently reason about things like this and be confident that pointers will always be non-null. Meaning there is a lot more room for doubt about whether it is appropriate to say a given pointer is non-null.

But here we know for sure that we wont ever have a null pointer in at least the guards. Ideally the main storage point would be a late init `NonNull`, (i could probably actually do this if i use some unsafe and MaybeUninit to initialize the struct, but i figured I'd avoid doing fancy stuff like that initially). Not to mention that we dont leak the fact that we use `NonNull` to the rest of gdext anyway.

Basically my argument is that this is a very self contained library with relatively easy to reason about safety conditions and we can be sure these pointers wont be null, therefore it is appropriate to use `NonNull`. Whereas gdext is a big and complicated to reason about library, with intricate dependencies and invariants. so it's a lot harder to be confident in `NonNull` being appropriate so generally speaking it's better to avoid it.

But if we could be sure that some pointers in a datastructure will never be null in gdext, then it'd be more appropriate to model them as `NonNull`.

Reply:
update: it is in fact possible to make `ptr` be `NonNull<T>` instead of `Option<NonNull<T>>` by using `MaybeUninit` to initialize it.

Reply:
I think we can discuss this endlessly, my point is simply "use the most straightforward representation" (following KISS and YAGNI). There's enough complexity in this new crate, and using localized idioms like this make code harder to understand overall.

I have nothing against `NonNull` itself in the API if it expresses "this pointer cannot be null". But you try to use `Option<NonNull<T>>` for something like `LateInit<NonNull<T>>` but the `Option` type doesn't have late-init guarantees. It's the same as `*mut T` with the additional invariant that the pointer is initially null and later not.

Imo comments on the field declaration express this more clearly than `Option<NonNull<T>>`, which is not an established convention for the late-init usage you have here. So most people will just wonder "why use `*mut T` with extra steps".


Let's definitely avoid `unsafe` here, it doesn't make anything simpler.